backend/src/apiserver/resource/resource_manager.go

```diff
@@ -2048,3 +2058,137 @@ func (r *ResourceManager) GetTask(taskId string) (*model.Task, error) {
 	}
 	return task, nil
 }
+
+// ensureSemaphoreConfigMap parses the pipeline manifest and creates semaphore ConfigMaps if needed
+func (r *ResourceManager) ensureSemaphoreConfigMap(ctx context.Context, namespace, manifest string) error {
+	// Parse the manifest to extract semaphore configuration
+	semaphoreKey, mutexName, err := r.extractSemaphoreConfig(manifest)
+	if err != nil {
+		glog.Warningf("Failed to parse semaphore configuration from manifest: %v", err)
+		return nil // Don't fail run creation for semaphore parsing issues
+	}
+
+	// Create ConfigMap for semaphore_key if present
+	if semaphoreKey != "" {
+		if err := r.semaphoreManager.EnsureSemaphoreConfigMap(ctx, namespace, semaphoreKey); err != nil {
+			glog.Errorf("Failed to ensure semaphore ConfigMap for key %s: %v", semaphoreKey, err)
+			return nil // Don't fail run creation for ConfigMap issues
+		}
+	}
+
+	// Create ConfigMap for mutex_name if present (same logic, treat as semaphore with limit 1)
+	if mutexName != "" {
+		if err := r.semaphoreManager.EnsureSemaphoreConfigMap(ctx, namespace, mutexName); err != nil {
+			glog.Errorf("Failed to ensure mutex ConfigMap for name %s: %v", mutexName, err)
+			return nil // Don't fail run creation for ConfigMap issues
+		}
+	}
+
+	return nil
+}
+
+// getStringFromYAML safely extracts string values from yaml.v2 maps that can be map[interface{}]interface{}
+func (r *ResourceManager) getStringFromYAML(data interface{}, keys ...string) (string, bool) {
+	current := data
+
+	for _, key := range keys {
+		switch v := current.(type) {
+		case map[string]interface{}:
+			var ok bool
+			current, ok = v[key]
+			if !ok {
+				return "", false
+			}
+		case map[interface{}]interface{}:
+			var ok bool
+			current, ok = v[key]
+			if !ok {
+				return "", false
+			}
+		default:
+			return "", false
+		}
+	}
+
+	// Final conversion to string
+	if str, ok := current.(string); ok {
+		return str, true
+	}
+	return "", false
+}
+
+// extractSemaphoreConfig parses the pipeline manifest to extract semaphore_key and mutex_name
+func (r *ResourceManager) extractSemaphoreConfig(manifest string) (semaphoreKey, mutexName string, err error) {
```

Comment:
This is duplicating a lot of code. Let's use `template.NewV2SpecTemplate` to parse the platform spec instead.